### PR TITLE
RES-2190: Fix labels set to -100 in finetuning tasks

### PR DESCRIPTION
**Pull request by benja-matic**
Regarding RES-2190. Looks like HuggingFace trainer looks at args.dataloader_drop_last for train, and eval loaders. Workaround is to turn flip that to False during evaluation in compute_metrics_task, flip it back to True at the end (if it was originally true). A few minor ignorable formatting edits will merge in as well. Finally, there's a finetuning experiment for tiny_bert50k.

### CODE REVIEW

File: projects/transformers/run_utils.py [hunk 1]
```diff
@@ -125,7 +125,22 @@ def evaluate_tasks(trainer, output_dir, tasks, eval_datasets):
...
+    if trainer.args.dataloader_drop_last:
+        drop_last=True
+        trainer.args.dataloader_drop_last=False
+        logging.info("Switched trainer.args.dataloader_drop_last to False for evaluation")
+
     for eval_dataset, task in zip(eval_datasets, tasks):
+        vals, cnts = np.unique(np.array(eval_dataset['label']), return_counts=True)
+        logging.info(f"Label distribution for {task} before calling trainer.evaluate")
+        for val in range(len(vals)):
+            logging.info(f"Label={vals[val]}: {cnts[val]} examples")
```
Comment (mvacaporale) [inline projects/transformers/run_utils.py:143]: Did you intentionally mean to keep this logging? Do you think we need it?

File: projects/transformers/run_utils.py [hunk 2]
```diff
@@ -727,41 +751,31 @@ def compute_metrics_task(ep: EvalPrediction, metric=None,
...
-        logging.info(f"Predictions: {Counter(preds).most_common()}")
-        logging.info(f"Labels: {Counter(ep.label_ids).most_common()}")
-
-    # Ignore the -100 labels - not able to tokenize?
-    # TODO: investigate why a few labels are -100 in all tasks
-    label_ids = ep.label_ids[np.where(ep.label_ids != -100)]
-    preds = preds[np.where(ep.label_ids != -100)]
-    logging.info(f"Removing {1-(len(label_ids) / len(ep.label_ids)):.2%} samples "
-                 "from evaluation set where label == -100")
+    assert -100 not in ep.label_ids, "unknown source of -100 labels"
```
Comment (mvacaporale) [inline projects/transformers/run_utils.py:756]: Nit pick: Could you fix the punctuation? "Unknown source of -100 labels."

File: projects/transformers/run_utils.py [hunk 3]
```diff
@@ -916,7 +930,7 @@ def model_init():
             "HP search with saved models not supported."
         logging.info("Pretraining new model from scratch")
 
-        # Instantiate model; possibly one of our custom sparse models.
+    # Instantiate model; possibly one of our custom sparse models.
```
Comment (mvacaporale) [inline projects/transformers/run_utils.py:935]: The indentation looks like it was changed here as well.

**Comment by mvacaporale** [comment]
Good find. The solution is simple and straightforward. Note, there could be others ways. For one, we could override `get_eval_dataloader`, but I think that would a bit cumbersome.

**Review by mvacaporale — APPROVED** [review]

File: projects/transformers/run.py [hunk 4]
```diff
@@ -299,12 +299,13 @@ def run_finetuning_single_task(
...
         "validation_matched" if data_args.task_name == "mnli" else "validation"
     ]
+
     test_dataset = None
-    if ((data_args.task_name is not None or data_args.test_file is not None)
-       and training_args.do_predict):
-        test_dataset = tokenized_datasets[
-            "test_matched" if data_args.task_name == "mnli" else "test"
-        ]
+    if (data_args.task_name is not None or data_args.test_file is not None):
```
Comment (lucasosouza) [inline projects/transformers/run.py:304]: I didn't understand this change, what is the difference between the old version and the new?

**Review by lucasosouza — CHANGES_REQUESTED** [review]
Also seems like a good solution to me, it is a hack but it should work fine and with no extra computation or memory required. I left some very minor comments, please take a look

File: projects/transformers/run_utils.py [hunk 5]
```diff
@@ -141,6 +153,12 @@ def evaluate_tasks(trainer, output_dir, tasks, eval_datasets):
                     logging.info(f"  {key} = {value}")
                     writer.write(f"{key} = {value}\n")
 
+    # if you want drop_last for training, this toggles it back on
+    if drop_last:
+        trainer.args.dataloader_drop_last = True
+        logging.info("Switched trainer.args.dataloader_drop_last"
```
Comment (lucasosouza) [inline projects/transformers/run_utils.py:159]: I think we should lower the level of this logging, logging.debug should be enough. There is no useful information to the user here, it is more of a print for debugging purposes.

**Review by lucasosouza — APPROVED** [review]

File: projects/transformers/run.py (hunk 4 above)
Comment (benja-matic) [reply projects/transformers/run.py:304]: No semantic difference. I just couldn't figure out how to make it pep compliant. I was getting issues about the indent, or the length of the line, so just broke it up.

File: projects/transformers/run_utils.py (hunk 5 above)
Comment (benja-matic) [reply projects/transformers/run_utils.py]: Got it. Those were mainly to verify that my fix was working properly, not needed going forward. I removed the bit above and a couple related ones.

**Comment by mvacaporale** [comment]
@benja-matic Do you know how this change affects the fine-tuning results? We should probably rerun those from the README (including `bert_1mi`, `bert_100k`, `sparse_80%_kd_onecycle_lr_rigl`, and `sparse_80%_kd_onecycle_lr`) @lucasosouza What do you think? Is this necessary? My concern is that it may make it harder to contextualize new results if we don't rerun previous fine-tuning experiments.

**Comment by benja-matic** [comment]
> @benja-matic Do you know how this change affects the fine-tuning results? We should probably rerun those from the README (including `bert_1mi`, `bert_100k`, `sparse_80%_kd_onecycle_lr_rigl`, and `sparse_80%_kd_onecycle_lr`) @lucasosouza What do you think? Is this necessary? My concern is that it may make it harder to contextualize new results if we don't rerun previous fine-tuning experiments. Or at the very least, we should rerun one or two just to make sure the results are negligibly affected.


@mvacaporale don't know just yet. I'm rerunning fine tuning on baseline models this afternoon.
